File: resume-saas-v4/lib/linkedin-skills.ts
/** Skill-only values for new results and previously saved LinkedIn analyses. */
const headings = new Set([
    "skill", "skills", "technical skills", "soft skills", "top skills",
    "certification", "certifications", "licenses & certifications", "contact",
    "contact info", "contact information", "personal information", "personal details",
    "email", "email address", "phone", "phone number", "mobile", "address",
    "name", "full name", "location", "education", "experience", "projects",
    "summary", "about", "interests", "references", "n/a", "none", "null",
]);
const skillGroups = new Set([
    "technical", "soft", "languages", "frameworks", "databases", "cloud", "devops",
    "other", "items", "values", "list", "skills", "skill", "technicalskills",
    "softskills", "topskills", "skilldetails", "associatedskills", "technologies",
    "technology", "techstack", "tools", "competencies", "expertise",
]);
type SkillRecord = Record<string, unknown>;
export type LinkedInSkill = { name: string; endorsements?: string; positions?: string[] };

export function cleanSkillName(value: unknown): string {
    if (typeof value !== "string") return "";
    const text = value.normalize("NFKC").replace(/[\u0000-\u001f\u007f\u200b-\u200f\ufeff]/g, "")
        .replace(/\s+/g, " ").replace(/^[\s*•:-]+|[\s*•:-]+$/g, "");
    if (!text || text.length > 100 || !/\p{L}/u.test(text)) return "";
    if (headings.has(text.toLowerCase()) || text.includes("@")) return "";
    if (/(?:https?:\/\/|www\.|mailto:|tel:|(?:linkedin\.com|github\.com)\/)/i.test(text)) return "";
    if (/^(?:e-?mail|phone|mobile|telephone|contact|address|location|full name)\s*[:=]/i.test(text)) return "";
    const phones = text.match(/(?<!\w)\+?\d[\d\s().-]{5,}\d(?!\w)/g) || [];
    if (phones.some(phone => (phone.match(/\d/g) || []).length >= 7)) return "";
    return text;
}

function splitSkills(text: string): string[] {
    let depth = 0, start = 0;
    const result: string[] = [];
    for (let i = 0; i < text.length; i++) {
        if ("([".includes(text[i])) depth++;
        else if (")]".includes(text[i])) depth = Math.max(0, depth - 1);
        else if (",;|\n\r•".includes(text[i]) && depth === 0) {
            result.push(text.slice(start, i));
            start = i + 1;
        }
    }
    result.push(text.slice(start));
    return result;
}

export function cleanSkills(value: unknown, limit = 80): string[] {
    const result: string[] = [], seen = new Set<string>();
    function add(item: unknown) {
        if (typeof item === "string") {
            for (const part of splitSkills(item)) {
                const name = cleanSkillName(part);
                if (name && !seen.has(name.toLowerCase()) && result.length < limit) {
                    seen.add(name.toLowerCase());
                    result.push(name);
                }
            }
        } else if (Array.isArray(item)) item.forEach(add);
        else if (item && typeof item === "object") {
            const record = item as SkillRecord;
            for (const field of ["name", "title", "skill", "text"]) {
                if (typeof record[field] === "string" && record[field].trim()) {
                    add(record[field]);
                    return;
                }
            }
            for (const [key, child] of Object.entries(record)) {
                if (skillGroups.has(key.toLowerCase().replace(/[^a-z]/g, ""))) add(child);
            }
        }
    }
    add(value);
    return result;
}

export function cleanSkillDetails(value: unknown): LinkedInSkill[] {
    const seen = new Set<string>();
    return (Array.isArray(value) ? value : cleanSkills(value)).flatMap(item =>
        cleanSkills(item).filter(name => {
            if (seen.has(name.toLowerCase())) return false;
            seen.add(name.toLowerCase());
            return true;
        }).map(name => ({ ...(item && typeof item === "object" ? item : {}), name }))
    );
}

/** Decode legacy JSON strings, then clean only skill fields; contact info stays intact. */
export function sanitizeLinkedInProfile(value: unknown): Record<string, any> | null {
    let parsed = value;
    for (let i = 0; i < 5 && typeof parsed === "string"; i++) {
        try { parsed = JSON.parse(parsed); } catch { return null; }
    }
    if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) return null;
    const profile = { ...(parsed as SkillRecord) };
    if ("skills" in profile) profile.skills = cleanSkills(profile.skills);
    if ("topSkills" in profile) profile.topSkills = cleanSkills(profile.topSkills);
    if ("skillDetails" in profile) profile.skillDetails = cleanSkillDetails(profile.skillDetails);
    for (const section of ["experience", "education", "projects"]) {
        if (!Array.isArray(profile[section])) continue;
        profile[section] = profile[section].map((item: unknown) => {
            if (!item || typeof item !== "object" || Array.isArray(item)) return item;
            const entry = { ...(item as SkillRecord) };
            if ("skills" in entry) entry.skills = cleanSkills(entry.skills);
            if ("associatedSkills" in entry) entry.associatedSkills = cleanSkills(entry.associatedSkills).join(", ");
            return entry;
        });
    }
    return profile;
}

File: Backend/Full_Backend/app/utils/linkedin_skills.py
"""Skill-only normalization for LinkedIn imports, rewrites and keyword enrichment."""
import re
import unicodedata

_HEADINGS = {
    "skill", "skills", "technical skills", "soft skills", "top skills",
    "certification", "certifications", "licenses & certifications", "contact",
    "contact info", "contact information", "personal information", "personal details",
    "email", "email address", "phone", "phone number", "mobile", "address",
    "name", "full name", "location", "education", "experience", "projects",
    "summary", "about", "interests", "references", "n/a", "none", "null",
}
_SKILL_FIELDS = {
    "skills", "skill", "technicalskills", "softskills", "topskills", "skilldetails",
    "associatedskills", "technologies", "technology", "techstack", "tools",
    "competencies", "expertise",
}
_CONTACT_FIELDS = {"contact", "contactinfo", "contactinformation", "personalinfo", "personaldetails", "email", "phone", "mobile", "address"}
_GROUP_FIELDS = {"technical", "soft", "languages", "frameworks", "databases", "cloud", "devops", "other", "items", "values", "list"}
_PHONE = re.compile(r"(?<!\w)\+?\d[\d\s().-]{5,}\d(?!\w)")
_LINK = re.compile(r"(?:https?://|www\.|mailto:|tel:|(?:linkedin\.com|github\.com)/)", re.I)
_CONTACT_LABEL = re.compile(r"^(?:e-?mail|phone|mobile|telephone|contact|address|location|full name)\s*[:=]", re.I)


def clean_skill_name(value) -> str:
    if not isinstance(value, str):
        return ""
    text = unicodedata.normalize("NFKC", value)
    text = re.sub(r"[\x00-\x1f\x7f\u200b-\u200f\ufeff]", "", text)
    text = re.sub(r"\s+", " ", text).strip(" -*•\t:")
    if not text or len(text) > 100 or not any(char.isalpha() for char in text):
        return ""
    if text.lower() in _HEADINGS or "@" in text or _LINK.search(text) or _CONTACT_LABEL.search(text):
        return ""
    if any(sum(char.isdigit() for char in match.group()) >= 7 for match in _PHONE.finditer(text)):
        return ""
    return text


def _key(value) -> str:
    return re.sub(r"[^a-z]", "", str(value).lower())


def _split_skills(text: str):
    # Keep CI/CD, C++, and parenthesized lists such as LLMs (GPT-4, Gemini).
    depth, start = 0, 0
    for index, char in enumerate(text):
        if char in "([":
            depth += 1
        elif char in ")]":
            depth = max(0, depth - 1)
        elif char in ",;|\n\r•" and depth == 0:
            yield text[start:index]
            start = index + 1
    yield text[start:]


def coerce_skill_list(value, limit: int = 80) -> list[str]:
    skills, seen = [], set()

    def add(item):
        if isinstance(item, str):
            for part in _split_skills(item):
                name = clean_skill_name(part)
                if name and name.lower() not in seen and len(skills) < limit:
                    seen.add(name.lower())
                    skills.append(name)
        elif isinstance(item, list):
            for child in item:
                add(child)
        elif isinstance(item, dict):
            # A skill object can carry endorsements, URLs or contact metadata;
            # those values must never become additional skills.
            for field in ("name", "title", "skill", "text"):
                if isinstance(item.get(field), str) and item[field].strip():
                    add(item[field])
                    return
            for field, child in item.items():
                if _key(field) in _GROUP_FIELDS | _SKILL_FIELDS:
                    add(child)

    add(value)
    return skills


def collect_master_skills(master_data, limit: int = 35) -> list[str]:
    candidates = []

    def walk(value):
        if isinstance(value, dict):
            for field, child in value.items():
                key = _key(field)
                if key in _CONTACT_FIELDS:
                    continue
                if key in _SKILL_FIELDS:
                    candidates.extend(coerce_skill_list(child))
                elif isinstance(child, (dict, list)):
                    walk(child)
        elif isinstance(value, list):
            for child in value:
                walk(child)

    walk(master_data)
    return coerce_skill_list(candidates, limit)


def clean_skill_details(value) -> list[dict]:
    result, seen = [], set()
    items = value if isinstance(value, list) else coerce_skill_list(value)
    for item in items:
        for name in coerce_skill_list(item):
            if name.lower() in seen:
                continue
            seen.add(name.lower())
            result.append({**(item if isinstance(item, dict) else {}), "name": name})
    return result
